### PR TITLE
Fix warning CA2002: Do not lock on objects with weak identity

### PR DESCRIPTION
**Pull request by zerhacken**
This commit changes 'lock (this)' usage into a lock on a private
object which only instance is knowledgeable about.

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs [hunk 1]
```diff
@@ -25,9 +25,11 @@ public SymbolDeclaredCompilationEvent(Compilation compilation, ISymbol symbol, L
         private Lazy<SemanticModel> _lazySemanticModel;
         private SemanticModel _semanticModel;
         private WeakReference<SemanticModel> _weakModel = null;
+        // lockable object only instance is knowledgeable about
+        private readonly object lockable = new object();
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs]: Could you use the name `gate` here instead?  I believe that is the verbiage we most commonly associate with `object` values used only in `lock` statements.

File: src/Scripting/Core/ScriptBuilder.cs [hunk 2]
```diff
@@ -49,6 +49,9 @@ internal class ScriptBuilder
         private readonly UncollectibleCodeManager _uncollectibleCodeManager;
         private readonly CollectibleCodeManager _collectibleCodeManager;
 
+        // lockable object only instance is knowledgeable about
+        private readonly object lockable = new object();
```
Comment (jaredpar) [inline src/Scripting/Core/ScriptBuilder.cs]: Same

File: src/Scripting/Core/ScriptBuilder.cs [hunk 3]
```diff
@@ -243,7 +246,10 @@ private sealed class CollectibleCodeManager : AssemblyLoader
             internal readonly AssemblyLoader assemblyLoader;
             private readonly AssemblyIdentity _dynamicAssemblyName;
 
-            // lock(this) on access
+            // lockable object only instance is knowledgeable about
+            private readonly object lockable = new object();
```
Comment (jaredpar) [inline src/Scripting/Core/ScriptBuilder.cs]: Please maintain the `lock(gate) on access` portion of the comment.

**Comment by jaredpar** [comment]
Thanks for doing this.  Left some minor comments on the change.  

I think this will be good to go once we are in a position to accept PR's (very, very soon).

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs (hunk 1 above)
Comment (mikedn) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs]: You mean `_gate` :smile:

**Comment by zerhacken** [comment]
Changed field name from "lockable" to "_gate" and fixed comment (maintained  "lock(gate) on access")

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs [hunk 4]
```diff
@@ -25,9 +25,11 @@ public SymbolDeclaredCompilationEvent(Compilation compilation, ISymbol symbol, L
         private Lazy<SemanticModel> _lazySemanticModel;
         private SemanticModel _semanticModel;
         private WeakReference<SemanticModel> _weakModel = null;
+        // lockable object only instance is knowledgeable about
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs]: Comment isn't really needed, but XML doc comments would be more appropriate if you're going to keep them. IDE goodness and all. Ditto for everything else.

**Comment by zerhacken** [comment]
@jasonmalinowski fixed comments to XML doc, if a commentless fix is preferred I will of course remove them as necessary. I would personally keep the comments as a fair warning.

**Comment by gafter** [comment]
:+1:

**Comment by zerhacken** [comment]
git fetch upstream
git rebase upstream/master

**Comment by srivatsn** [comment]
@zerhacken - If you haven't already, can you please sign the CLA here - https://cla2.dotnetfoundation.org ?

**Pull request merged**